File: scripts/geo.py
#!/usr/bin/env python3
"""Great-circle distance between two (lat, lng) points, in kilometres.

The one interface for this in the codebase's Python side. Before this module,
the same haversine formula was pasted independently six times in scripts/
(with three drifted call signatures and two drifted units), plus four more
times on the TypeScript side (see site/src/lib/walks.ts::haversineKm for that
counterpart). Deepening candidate #1 from the 2026-08-13 architecture review:
https://github.com/burgmanshidde-ancienttrees/ancienttrees/issues/1
"""
import math

EARTH_RADIUS_KM = 6371.0


def km(a, b):
    """Distance between two (lat, lng) points, in kilometres."""
    lat1, lat2 = math.radians(a[0]), math.radians(b[0])
    dlat = math.radians(b[0] - a[0])
    dlng = math.radians(b[1] - a[1])
    h = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlng / 2) ** 2
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(h))

File: ios/AncientTreesKit/Sources/Geo.swift
// Distance on the sphere. One implementation, because the Python side learned
// this lesson the expensive way: the same haversine was pasted six times in
// scripts/ with three drifted call signatures and two drifted units before
// scripts/geo.py took over. This is that module for the app.

import Foundation

public enum Geo {
    public static let earthRadiusKm = 6371.0

    /// Great-circle distance in kilometres.
    public static func km(_ a: (lat: Double, lng: Double),
                          _ b: (lat: Double, lng: Double)) -> Double {
        let dLat = (b.lat - a.lat) * .pi / 180
        let dLng = (b.lng - a.lng) * .pi / 180
        let lat1 = a.lat * .pi / 180
        let lat2 = b.lat * .pi / 180
        let h = sin(dLat / 2) * sin(dLat / 2)
              + sin(dLng / 2) * sin(dLng / 2) * cos(lat1) * cos(lat2)
        return 2 * earthRadiusKm * asin(min(1, sqrt(h)))
    }

    /// Walking minutes at 4.5 km/h, the pace the website's walks already assume.
    public static func walkingMinutes(km: Double) -> Int {
        Int((km / 4.5 * 60).rounded())
    }
}

public extension Tree {
    func distanceKm(from lat: Double, _ lng: Double) -> Double {
        Geo.km((lat: lat, lng: lng), (lat: self.lat, lng: self.lng))
    }
}
